src/algebra/order/monoid_lemmas_gt_zero.lean

```diff
@@ -0,0 +1,131 @@
+
+*  `sx : α>0 → α → α` is defined as `sx a b = a * b`, with `a` coerced to `α` by virtue of being in
+  a subtype of `α`;
+*  `dx : α>0 → α → α` is defined as `dx a b = b * a`, with `a` coerced to `α` by virtue of being in
+  a subtype of `α`.
+
+We combine this with (`contravariant_`) `covariant_class`es to assume that multiplication by
+positive elements is (strictly) monotone on a `mul_zero_class`, `monoid_with_zero`,...
+More specifically, we use extensively the following typeclasses:
+
+* monotone left
+* * `covariant_class α>0 α sx (≤)`, expressing that multiplication by positive elements on the left
+    is monotone;
+* * `covariant_class α>0 α sx (<)`, expressing that multiplication by positive elements on the left
+    is strictly monotone;
+* monotone right
+* * `covariant_class α>0 α dx (≤)`, expressing that multiplication by positive elements on the right
+    is monotone;
+* * `covariant_class α>0 α dx (<)`, expressing that multiplication by positive elements on the right
+    is strictly monotone.
+* reverse monotone left
+* * `contravariant_class α>0 α sx (≤)`, expressing that multiplication by positive elements on the
+    left is reverse monotone;
+* * `contravariant_class α>0 α sx (<)`, expressing that multiplication by positive elements on the
+    left is strictly reverse monotone;
+* reverse reverse monotone right
+* * `contravariant_class α>0 α dx (≤)`, expressing that multiplication by positive elements on the
+    right is reverse monotone;
+* * `contravariant_class α>0 α dx (<)`, expressing that multiplication by positive elements on the
+    right is strictly reverse monotone.
+
+##  Formalization comments
+
+We use `α>0 = {x : α // 0 < x}` with a strict inequality since in most cases what happens with `0`
+is clear.  This creates a few bumps in the first couple of proofs, where we have to split cases on
+whether an element is `0` or not, but goes smoothly after that.  A further advantage is that we
+only introduce notation for the positive elements and we do not need also the non-negative ones.
+-/
+
+/- I am changing the file `algebra/order/monoid_lemmas` incrementally, with the idea of
+reproducing almost all of the proofs in `algebra/order/ring` with weaker assumptions. -/
+
+universe u
+variable {α : Type u}
+
+/--  The type of positive elements in a type that has a `0` and a `<`. -/
+@[reducible]
+def gt_zero (α : Type u) [has_zero α] [has_lt α] := {x : α // 0 < x}
+
+/-  Notation for positive elements
+https://
+leanprover.zulipchat.com/#narrow/stream/113488-general/topic/notation.20for.20positive.20elements
+-/
+local notation `α>0` := gt_zero α
+
+namespace gt_zero
+
+/--  `sx` is the multiplication of an element of the subtype `α>0 = {x : α // 0 < x}` of positive
+elements by an element of the type itself.  The element of the subtype appears on the left:
+`sx a b = a * b`.
+
+`dx` is the multiplication in the other order. -/
+def sx [has_zero α] [has_lt α] [has_mul α] : α>0 → α → α :=
```

Comment:
Are these for "internal use" only? As an aid in proving the lemmas below? Should they be `private abbreviation`s?

Reply:
These definitions don't look great (if these are used in defs or instances, you will need to rewrite them and have simp lemmas...). Can they be spelled as `(x : α) * y` and `x * (y : α)` instead of introducing new definitions?

Reply:
Those definitions were my way of getting Lean to understand succintly what I meant.  There are probably better solutions, but I tried a few things and this is the best I could find!

Below is my attempt at not using `sx`.  If you can golf them, I would be very happy!

I do not really know the implications of `private abbreviation`s.  I do not like `private` too much, since if I found them useful, someone else might as well and sealing them off seems counterproductive.  I am not strongly against it, though!

```lean
lemma mul_lt_mul_left' [covariant_class α>0 α (λ x y, x * y) (<)]
  {a b c : α} (bc : b < c) (a0 : 0 < a) :
  a * b < a * c :=
let a₀ : α>0 := ⟨a, a0⟩ in
suffices (λ (x : α>0) (y : α), (x : α) * y) a₀ b < (λ (x : α>0) (y : α), (x : α) * y) a₀ c,
by simpa using this,
covariant_class.elim a₀ bc
```

Reply:
If you add an `covariant_class.elim'` with the `covariant_class` argument explicit, I'm almost certain that `h.elim' ⟨a, a0⟩ bc` is also a valid proof...

Reply:
Ok, I personally do not like too much having to use `@`, but can live with doing it every once in a while.  Thus, if it is only to get the first 4 proofs started and then they disappear, I would prefer not to add the `elim'`.

However, looking at it again, I personally find it quite inconvenient to type `covariant_class α>0 α (λ x y, x * y) (<)` *every* time I want an assumption "multiplication by positive elements is monotone".  I am starting to wonder whether to give `sx` and `dx` better names, proving a couple of `simp` lemmas about them and continue using them.

Would this be acceptable?  I find `(λ x y, x * y)` a really obfuscated version of "multiply on the left!

Reply:
Would it be better received if I called `pos_mul` and `mul_pos` (in the `zero_lt` namespace) what is currently called `sx` and `dx`?

Reply:
Or also, what would you think of what is below? (Plus similar abbreviations for `pos_mul_mono`, `mul_pos_strict_mono` and `mul_pos_mono`.)

```lean
abbreviation pos_mul_strict_mono (α : Type*) [has_mul α] [has_zero α] [has_lt α] : Prop :=
covariant_class {x : α // 0 < x} α (λ x y, x * y) (<)
```

In terms of reading and typing this, I find this much nicer than anything else that has been suggested so far!  With this, even the local notation can go!

Reply:
> Would this be acceptable? I find `(λ x y, x * y)` a really obfuscated version of "multiply on the left!

I think `(λ x y, x * y)` is *a lot* clearer than any `sx` or `pos_mul` you can write.

Try looking at it from the perspective of an experienced Lean user that hasn't used this file before. 
Which of `(λ x y, x * y)`, `sx`, `pos_mul` do you think they are most likely to understand?

Reply:
I don't quite understand what your problem with `covariant_class α>0 α (λ x y, x * y) (<)` is. Is it the number of keystrokes? The difficulty in understanding what it means?

The first is improved a bit my adding new abbreviations. However, for the second: notice that every abbreviation you add is one more thing for the user to learn to use this library.
I'm okay with adding abbreviations for the classes (though I'd prefer not to), but I still think we shouldn't add the definitions for "multiply a positive element with an element".